packages/flutter_test/lib/src/controller.dart

```diff
@@ -23,6 +23,201 @@ const double kDragSlopDefault = 20.0;
 
 const String _defaultPlatform = kIsWeb ? 'web' : 'android';
 
+/// Class that programatically interacts with the [Semantics] tree.
+///
+/// Allows for testing of the [Semantics] tree, which is used by accessibility
+/// tools, search engines, and other analysis software to determine the meaning
+/// of an application.
+///
+/// Should be accessed through [WidgetController.semantics]. If no custom
+/// implementation is provided, a default [SemanticsController] will be created.
+class SemanticsController {
```

Comment:
I am a little bit concern about adding a brand new interface just for retrieving semantics node traversal order which can be in a matcher instead. It feels to me that we are diverging from the previous pattern.

Reply:
While this currently is just for traversal and a new place for the `getSemantics` functionality, I plan on extending it further with #112413.

Could you explain how the traversal can be a `Matcher`? I looked into the `Matcher` and `Finder` interfaces. `Finder` is based on `Elements`, which makes it incompatible with `SemanticsNode`s. `Matcher` is used for validations, not for getting something, so what would you suggest gets the thing that is matched against?

Reply:
I would agree that to have the single new method, it wouldn't make sense to create a whole new interface for it. But with #112413 requiring more new semantics related methods, I took this as an opportunity to make a new place for easier discoverability for semantics based tests.

It also allows us to implement #112413 with a cleaner (if more verbose) API if we want to. Instead of a single `performAction` method, we could create individual `tap`, `scrollUp`, etc. methods if we want to without having to add the word `Semantics` in the method name for each one.

Reply:
if you have element you can also get the semantics node on it i think? I was expecting something like
```
expect(find.byType(ListView), hasChildSemanticsOrder(<finder>(find.byText('item 0'), find.byText('item 1'))))
```
The perform actions issue does seem like a good idea to have this interface though.

Reply:
in response to https://github.com/flutter/flutter/pull/113133#discussion_r992706807,

I meant I prefer to keep the WidgetTester.getSemantics over the WidgetTexter.semantics.find if we can keep SemanticsController private. I still leaning slightly toward including every api in WidgetTester and keep SemanticsController private even for perform actions though.

Reply:
I was originally thinking something along the lines of what you suggested for the finder/matcher combo as well, but it ended up being clunky at best. The three major issues I ran into that made me shift to what's here now were these:

1. While you _can_ get the semantics node of a particular widget, it can be a bit clunky to find the right one. To start with, if the Element doesn't own it's own node, you actually get the node for the closest ancestor _containing_ that node. So in the example you gave, you'd actually get something above this ListView. So if you had a Scaffold there, you would also get the FloatingActionButton.
2. Related to #1, but having a matcher that uses finders is not very useful for Semantics. Hence why I originally went with a matcher that takes a list of matchers, but it turns out that already exists with the `orderedEquals` and `containsAllInOrder` matchers.
3. Lastly, picking a place in the middle of the tree to start the traversal isn't easy. While everything is _built_ like a tree, it's _interacted with_ like a list of nodes. You swipe or tab and a different node is focused, if you reach a leaf of the tree, the navigation moves to the next branch automatically, you don't have to manually go back up the tree. What I ran into is trying to start the depth first search at one part and continue the search from there meant that I was getting funky orders when trying to navigate from the leaf of one branch into a different branch of the tree. As implemented, the `simulatedAccessibilityTraversal` _looks_ like you start and end in the middle of that list (even though it technically does the depth first search in total) and lets the tester interact with that traversal like they would navigate the actual app. You can't cleanly do the full traversal without either requiring that the root node is always passed in, or without access to the `binding`.

Reply:
Do you have specific reasons for why you prefer including all of the semantics APIs in WidgetTester? I went with the SemanticsController and `semantics` property for the following reasons:

* Discoverability. If you start typing "semantics" you'll get a single property that you can get type "." and get full access to any supported semantics testing tools we have.
* Maintainability. Semantics is basically a parallel tree, and everything in WidgetController is aimed at the Element tree, and a bit of the render tree. With SemanticsController separated out, it allows us to have a separate controller with a single responsibility: "interact with the semantics tree for testing". It lets us overlap names where things may already exist on the `WidgetController`, such as the renamed `find` method, or potentially something like a `any` method later on.

Reply:
I have several reasons I prefer including all API in widget controller, but none of them is a must. I am 
1. OOPs general suggest interface abstraction and void accessing instance property to perform operations directly. Flutter does not really need to follow this rule though. Take the Widgettester.binding for one example, WidgetTester wraps it with async guard to implement the pump method. We would probably need to do the same thing if we provide async api in SemanticsController, and it will be non trivial to support if people accessing the SemanticsController directly.
2. As you said WidgetTester can access both element tree and rendering tree, why semantics as separate API? We should unify it, whether they are all separated sets of API or all in one. Separating them all seems a bit verbose to me.


> So in the example you gave, you'd actually get something above this ListView. So if you had a Scaffold there, you would also get the FloatingActionButton.

which is fine right? as long as the semantics node created beneath the widget is a descendant of that node 

> hat I ran into is trying to start the depth first search at one part and continue the search from there meant that I was getting funky orders when trying to navigate from the leaf of one branch into a different branch of the tree.

I am not sure if i understand, I thought even though it may skips unfocusable node, the order should still follow the depth-first search order, no? Do you have an example to illustrate the issue you are running into?

Reply:
I'm going to drop the preferences on the `semantics` property part of the conversation for now and we can revisit it later. The replies are getting a bit tough to respond to with both topics going in this one thread.

> > So in the example you gave, you'd actually get something above this ListView. So if you had a Scaffold there, you would also get the FloatingActionButton.
> 
> which is fine right? as long as the semantics node created beneath the widget is a descendant of that node

This actually would _not_ be fine IMO. If you are writing a test that checks that your list has three cards in it, and the test comes back with three cards and a FloatingActionButton, it'd be very confusing as to why my list has a FloatingActionButton in it at all.

> > What I ran into is trying to start the depth first search at one part and continue the search from there meant that I was getting funky orders when trying to navigate from the leaf of one branch into a different branch of the tree.
> 
> I am not sure if i understand, I thought even though it may skips unfocusable node, the order should still follow the depth-first search order, no? Do you have an example to illustrate the issue you are running into?

What I was running into was something like this: Given the tree below, we wanted to start at Node3. I would expect (from using the accessibility service on a device) that starting at Node3 would give me a traversal of [Node3, Node4, Node5].

```
Node1
|
--- Node2
|     |
|     --- Node3
|           |
|           --- Node4
|
--- Node5      
```

I'd start by getting Node3 and beginning my traversal. Since I didn't start at the root, I now have to eventually navigate _up_ the tree instead of just recursively going down the children, so when I add everything that's not visited, I end up getting: [Node3, Node4, Node2, Node1, Node5].

When I just traverse the children recursively from the root, I can make the full traversal and just cut off everything before Node3, but with this wonky traversal, I can't easily separate things out.

I _could_ start by finding the root of the tree by navigating up the parents until there isn't a parent, but instead I went with using the `binding` to directly get the root from the controller.

Alternatively, I could've tried (and did try, originally) to do the traversal in a `Matcher` with the "find the root first" solution, but the `SemanticsController` is much easier to work with, and lets us use that same traversal and pass it to several `Matchers` instead of having to implement the traversal in each new `Matcher` we need or having to create a new base class for those `Matcher`s.

The general idea of "get me this list of things" fit a lot better with a controller than it did with a `Matcher`. A `Finder` would have been a good fit in theory, but `SemanticsNode`s aren't `Element`s, so the `Finder` API is incompatible with the entire semantics tree.

Reply:
> Three cards and a FloatingActionButton, it'd be very confusing as to why my list has a FloatingActionButton in it at all.

Is the plan to do exact match?

If not, have additional nodes would still be fine right? and the customer only care about the match of the given semantics node finder.
If yes, how do we handle intermediate nodes that may or may not be focusable? As a user, how would one constuct the `fullTraversalMatchers` if some of the widget generate intermediate nodes?

Also, wouldn't we run into the same situation when the start of simulatedAccessibilityTraversal gets merge up to the parent that with a bunch of node from sibling tree?

> I would expect (from using the accessibility service on a device) that starting at Node3 would give me a traversal of [Node3, Node4, Node5].

In that case I would expect them to write something like
```
expect(Node1, hasChildSemanticsOrder([Node3, Node4, Node5]))
```
We can make it throw if the node1 doesn't contain all the child node in the subtree. To validate the order, we can treat the list of dfs children as circular. Although talkback may go through several nodes in the sibling trees until it circle back to the beginning, but the order should still hold.

Reply:
So, as written right now, the code solves all of the problems that you've raised, provides more functionality and flexibility, and does so in a way that it's reasonably simple to maintain and to use:

1. A controller (could be WidgetController or SemanticsController, we'll revisit that eventually) returns a list of nodes that simulates an accessibility traversal. This is done with a basic depth first traversal from the root (grabbed from the binding) that minimizes weird edge cases when attempting a bi-directional traversal. It also exposes a `start` and `end` that allows the user to get a sub-set of the traversal using any `Finder` they would like.
2. Since the controller returns a list of SemanticNodes that can be trimmed to a subset of the traversal, we can support both exact matching (with `orderedEquals`) AND fuzzy matching (with `containsAllInOrder`). Other matchers may also be useful here, such as the `contains`, `isNot`, or other iterable matchers.
3. The end user experience for writing a test is more natural. It does not require the test writer to understand how the semantics tree is structured, it only requires that they be able to run the app with accessibility enabled and to be able to note the order in which the items are read out.

At this point, this conversation is just me trying to justify my decisions, but I don't know what your concerns about using a controller to grab the semantics actually _are_. Could you clarify what problems having a controller grab the entire traversal list instead of creating a matcher that traverses from a parent are so I can try to address them?

Reply:
Awesome, glad to hear I was able to address those concerns!

Would you like to revisit the `semantics` property, or is that acceptable as well? I know we put that one to the side earlier, so I want to make sure we can address those concerns as well.

Reply:
sure, I mostly just concern about we are moving away from current style where WidgetTester try to provide all the API directly in its interface. One good thing about that is we can limit TestAsyncUtils.guard in WidgetTester class only. If the semanticscontroller will never provide an async API, then I think this is probably fine.

Reply:
What's the concern with using the `TestAsyncUtils` outside of WidgetTester/Controller? Now that you mention it, I _should_ probably put the `syncGuard` on the methods that are currently in `SemanticsController`.

Reply:
It looks like we already using it outside of the WidgetController, so i guess it probably doesn't matter then.

My original concern was that it may make it harder to compose the classes because we don't know if the class API will have async guard or not, and whether  the caller should apply their own async guard or not. Or if the SemanticsController is used in the context where async guard is not needed (for example non user code), it will still apply async guard unnecessary. It will be easier to maintain if only one class handles the async guard where all the other api go through it.

Reply:
To summarize several things in the discussion regards SemanticsController.

1. SemanticsController, I am totally up for create this class since it is a nice abstraction for functionality

3. WidgetTester.semantics as public getter and remove semantics related API from WidgetTester, this is the part that i had concerns with. If you look at TestWidgetsFlutterBinding class, you can see some of the API use async guard and some of them rely on the caller to use async guard. I think we are in a weird situation where some of its API needs to be wrapped by WidgetTester, while some other can be called directly from user code. I slightly prefer everything wrapped by WidgetTester since it may be easier to maintain. but I am actually ok either way. I will let you decide which direction we want to go.

Reply:
Thanks for summarizing, it really helps me understand where you're coming from.

Per the concerns about `TestAsyncUtils`, it looks like the concern is mostly about things that _nest_ those calls leading to ambiguity of who should actually be making this call. I agree that the current setup with `WidgetController.binding` is a bit of a mess in that regard...

That said, I think that the `WidgetController.semantics` property is a different enough use case that it won't run into the same issues there. The intent is to use that property to group semantics testing related functionality in a way that makes it more discoverable, whereas the `binding` property should really only be accessed _internally_ to `WidgetController` and its' implementations.

My biggest goals that drove me to expose that part of the API through a property were discoverability and extensibility. I want users to be able to easily find semantics testing tools when they're writing semantics tests. I also want to incentivize framework developers (myself included) to create a clean api specifically for semantics testing. That's harder when you have to differentiate semantics testing from element testing in the same class.

With those goals in mind, and my understanding that the `TestAsyncUtils` is at least _less_ likely for this particular property, I'm going to keep this part of the API segmented into the `semantics` property.

Thanks for digging deep into this @chunhtai! I appreciate the insight, and also for pointing out things for me to keep an eye on in future work.